### DISCUSSION TITLE
Efficient calculation of boundary misorientation statistics

### DISCUSSION THREAD

**Discussion by mfuhr-mpg** [mtex-toolbox/mtex#1288] [Ask Anything]
Hello,

I am trying to calculate the fraction of low and high angle grain boundary segments for each grain of an orientation map acquired using EBSD. The reason I want to do this calculation separately for each grain is that in the end I want to be able to compute a mean value for the fraction of high and low angle grain boundaries that uses the grain size as a weighting factor.
My initial approach was the following:

for i = 1:max(grains.id) %loop over all grains using their id as a counter
   
    G  = grains(grains.id==i);   % choose one grain
    GB = G.boundary('w','w'); % select the grain boundaries between W and W
    
    % case I:  segment misorientation <15° -> LAGB    
    isLAGB_w = angle(GB.misorientation)<=15*degree;
    % case II: segment misorientation >15° -> HAGB
    isHAGB_w = angle(GB.misorientation)>15*degree;
    
    % Divide the segments
    LAGB = GB(isLAGB_w);
    HAGB = GB(isHAGB_w);
   
    % Calculate the LAGB fraction
    Prop_LAGB = length(LAGB)/(length(HAGB)+length(LAGB));
    % Store the result in a separate variable
    Res(i,1) = G.grainSize;
    Res(i,2) = Prop_LAGB;
    
end

I am aware that this approach is quite inefficient, especially for a large number of grains. Thus, I was wondering if there is an opportunity to get the result I am after without calling each grain individually. Maybe one could store the result as a subgroup of grains.prop.

Thanks in advance for any help!

Best regards
Max Fuhr

**Comment by kilir** [comment]
HI,
if you pull the latest changes (0027b2abeddfc729cfa41c4a054f93e69dc65428 and b998fefa1d1238822e71d99ecc2aef284128b107) I think the following will do. Can you try this?
```
mtexdata ferrite

lagb = 2*degree;
hagb = 10*degree;
grains = calcGrains(ebsd,'angle',[hagb lagb])

% since innerBoundary might have also boundaries which have higher
% angles use a separate threshold
cond = (grains.innerBoundary.misorientation.angle < 8*degree)

%% number of segments
nhagb = grains.boundarySize;

nlagb = grains.subBoundarySize(cond)

plot(grains,nlagb./(nhagb+nlagb))


%% length of segments
lhagb = grains.boundary.segLength;

llagb = grains.subBoundaryLength(cond)

plot(grains,nlagb./(nhagb+nlagb))
```
Cheers,
Rüdiger

**Comment by mfuhr-mpg** [comment]
Hi Rüdiger,

thank you very much for the effort.
The code you sent (in combination with the changes to the functions) work well and in addition way faster than my approach.
I would have some more general questions to the approach you chose:
1) What is the outcome of defining two angular thresholds (in your example 2 and 8 degrees) for the calcGrains algorithm? If I execute the command,  I still "only" get an "ordinary" grains variable in which I don´t see the effect of the two threshold values.
2) How would one set the threshold angles of HAGB (misorientation angle >15 degrees) and LAGB (misorientation angle >2 & misorientation angle <15) in the example code you elaborated?
3) I assume there is a copy & paste error in your "%% length of segments" section: The plot command plots the segment number ratio. Adjusting the calculation of the ratio that is based on the length of the segments (llagb./(lhagb+llagb)), there seems to be a mismatch of matrix dimensions. Do I overlook something trivial here that explains why it is not working? 

Thanks in advance for your answers!
Best
Max

**Reply by kilir** [reply]
Hi,
wrt 1/2: the lagb threshold sets the lower limit for innerBoundary
```
mtexdata ferrite
grains = ebsd.calcGrains('angle',[2 15]*degree)
gb = grains.boundary('f','f')
plot(gb,gb.misorientation.angle/degree)
nextAxis
gi = grains('f').innerBoundary;
plot(gi,gi.misorientation.angle/degree)
mtexColorbar
```
wrt 3: yes of course, you are right,  the perniciousness of c&p

Cheers,
Rüdiger

**Reply by mfuhr-mpg** [reply]
Hi,

alright, thanks. This means that I would have to adjust `cond` in a way like
`cond = (grains.innerBoundary.misorientation.angle < 13*degree)`
in order to get the desired separation between LAGB and HAGB with misorientation angles like mentioned before?
Unfortunately there´s no resolution to the matrix mismatch I adressed in my third question. Sorry to annoy you but do you have an idea what might be wrong?

Best
Max

**Reply by kilir** [reply]
Yes, exactly, innerBoundary may contain boundaries with larger angles.

To compare the fractions derived by the length of segments the following should be correct:
```
lhagb = grains.perimeter;
llagb = grains.subBoundaryLength(cond)
plot(grains,llagb./(lhagb+llagb))
```
For non-smoothed boundaries it should give an identical picture than the one derived from the number of segments.

Cheers,
Rüdiger
